File: python/reorder-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reorderList(self, head):
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        prev = None
        next = None
        while slow:
            next = slow.next
            slow.next = prev
            prev = slow
            slow = next
        # return prev
        while prev.next:
            q = head.next
            w = prev.next
            head.next = prev
            prev.next = q
            head = q
            prev = w

File: java/reorder-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        ListNode prev = null;
        ListNode next = null;
        while (slow != null) {
            next = slow.next;
            slow.next = prev;
            prev = slow;
            slow = next;
        }
        // return prev
        ListNode p1 = head;
        ListNode p2 = prev;
        while (p2.next != null) {
            ListNode q = p1.next;
            ListNode w = p2.next;
            p1.next = p2;
            p2.next = q;
            p1 = q;
            p2 = w;
        }
    }
}
